### PR TITLE
[bugfix, enhancement] enable proper GPU offloading with fp64 support when dpctl unavailable 

### PR DESCRIPTION
**Pull request by icfaust**
## Description

This corrects a circular import issue with ```onedal.utils.validation``` when trying to create interfaces to the onedal backend in that file, the ```_device_offload``` imports from validation, and trying to use policy in the validation file will make a loop.  By moving the check to C++, it creates a better interface, and removes the need for the ```_device_offload``` import in ```_policy.py``` entirely.  

Major changes:
- Add sycl.cpp which contains SyclDevice and SyclQueue objects
- Remove device_lookup
- Move all SYCL-specific code from policy_common.hpp/.cpp to sycl_interfaces.hpp/.cpp
- Move pack_queue into sycl_interfaces.hpp/.cpp
- Move policy_common.hpp policy-specific functions into policy.cpp and add a policy.hpp header
- Remove DummySyclQueue
- Change _policy.py not to use DummySyclQueue
- Add ability to read sycl::queues from python integers via the PyLong method (used by pytorch in their XPUstreams to expose sycl_queue attribute)
- Fix importing from a py::capsule into sycl::queues
- Add ability to create and read SyclDevices ids from integers (will be useful for dlpack support)
- Add testing for compliance to dpctl's SyclQueue
- Add filter_string generation for diagnostic use/ conversion to dpctl SyclQueues

Future development with dlpack will be to have Enums from dlpack (see this PR: https://github.com/intel/scikit-learn-intelex/pull/1568)

---

PR should start as a draft, then move to ready for review state after CI is passed and all applicable checkboxes are closed.
This approach ensures that reviewers don't spend extra time asking for regular requirements.

You can remove a checkbox as not applicable only if it doesn't relate to this PR in any way.
For example, PR with docs update doesn't require checkboxes for performance while PR with any change in actual code should have checkboxes and justify how this code change is expected to affect performance (or justification should be self-evident).

Checklist to comply with **before moving PR from draft**:

**PR completeness and readability**

- [x] I have reviewed my changes thoroughly before submitting this pull request.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation to reflect the changes or created a separate PR with update and provided its number in the description, if necessary.
- [x] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/intel/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.
- [x] I have added a respective label(s) to PR if I have a permission for that.
- [x] I have resolved any merge conflicts that might occur with the base branch.

**Testing**

- [x] I have run it locally and tested the changes extensively.
- [x] All CI jobs are green or I have provided justification why they aren't.
- [x] I have extended testing suite if new functionality was introduced in this PR.

**Performance**

- [x] I have measured performance for affected algorithms using [scikit-learn_bench](https://github.com/IntelPython/scikit-learn_bench) and provided at least summary table with measured data, if performance change is expected.
- [x] I have provided justification why performance has changed or why changes are not expected.
- [x] I have provided justification why quality metrics have changed or why changes are not expected.
- [x] I have extended benchmarking suite and provided corresponding scikit-learn_bench PR if new measurable functionality was introduced in this PR.

### CODE REVIEW

**Comment by icfaust** [comment]
Rerun: http://intel-ci.intel.com/ef9fb16c-f83c-f13a-bcb3-a4bf010d0e2d

**Comment by icfaust** [comment]
Note to reviewers: we must figure out what to do with the deselections before merging this PR.  Some of these issues are only 32 bit GPU relevant.

**Comment by samir-nasibli** [comment]
> Note to reviewers: we must figure out what to do with the deselections before merging this PR. Some of these issues are only 32 bit GPU relevant.

Could you please bring more details about the issue?

File: onedal/common/sycl.cpp [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+void instantiate_sycl_interfaces(py::module& m){
+    py::class_<sycl::queue> syclqueue(m, "SyclQueue");
+    syclqueue.def(py::init<const sycl::device&>())
+        .def(py::init([](const std::string& filter) {
+                return get_queue_by_filter_string(filter);
+            })
+        )
+        .def(py::init([](const py::int_& obj) {
+                return get_queue_by_pylong_pointer(obj);
+            })
```
Comment (samir-nasibli) [inline onedal/common/sycl.cpp:35]: Could you please share the case when it is needed? Does it covered by tests?

File: onedal/common/sycl_interfaces.cpp [hunk 2]
```diff
@@ -70,18 +107,11 @@ sycl::queue get_queue_by_get_capsule(const py::object& syclobj) {
...
-    }
-    else if (py::isinstance(syclobj, pycapsule)) {
-        const auto caps = syclobj.cast<py::capsule>();
-        return extract_from_capsule(std::move(caps));
-    }
-    else {
-        throw std::runtime_error("Unable to interpret \"syclobj\"");
-    }
+sycl::queue get_queue_by_pylong_pointer(const py::int_& syclobj) {
+    //
```
Comment (samir-nasibli) [inline onedal/common/sycl_interfaces.cpp:111]: Probably here should be non-empty comment :)

File: onedal/dal.cpp [hunk 3]
```diff
@@ -102,6 +106,7 @@ namespace oneapi::dal::python {
 #else
     #ifdef ONEDAL_DATA_PARALLEL
     PYBIND11_MODULE(_onedal_py_dpc, m) {
+        init_sycl(m);
```
Comment (samir-nasibli) [inline onedal/dal.cpp:109]: Should not this be `sycl_interfaces` ?

File: onedal/common/sycl.cpp (hunk 1 above)
Comment (icfaust) [reply onedal/common/sycl.cpp:35]: https://github.com/pytorch/pytorch/blob/main/torch/csrc/xpu/Stream.cpp#L72

File: onedal/dal.cpp (hunk 3 above)
Comment (icfaust) [reply onedal/dal.cpp:109]: will change

**Comment by icfaust** [comment]
> > Note to reviewers: we must figure out what to do with the deselections before merging this PR. Some of these issues are only 32 bit GPU relevant.
> 
> Could you please bring more details about the issue?

Intel Max GPUs support fp64 computation. The dummy sycl queue assumes all gpus cannot compute using doubles.  The dummy sycl queue is used when DPCtl is not installed, which is the case in testing sklearn conformance with GPU.  This down conversion from double to float is causing results to be less precise and leading to many of the deselected gpu tests. I think there will be a follow-up PR which will differentiate GPU deselections based on hardware characteristics, at the moment I will leave all deselections.

File: onedal/dal.cpp (hunk 3 above)
Comment (icfaust) [reply onedal/dal.cpp:109]: @samir-nasibli would I need to rename sycl.cpp to sycl_interfaces.cpp, and if so, do you want me to move the contents of sycl.cpp to sycl_interfaces.cpp?

File: onedal/common/sycl.cpp (hunk 1 above)
Comment (Vika-F) [reply onedal/common/sycl.cpp:35]: Please add a comment somewhere that this is needed to accept pytorch tensors.
I think the better place for the comment is along with the function's definition. But it's up to you to decide about the comment's placement.

File: onedal/common/sycl.cpp [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+#include "onedal/common/pybind11_helpers.hpp"
+
+namespace py = pybind11;
+
+namespace oneapi::dal::python {
+
+#ifdef ONEDAL_DATA_PARALLEL
+
+void instantiate_sycl_interfaces(py::module& m){
+    py::class_<sycl::queue> syclqueue(m, "SyclQueue");
```
Comment (Vika-F) [inline onedal/common/sycl.cpp:27]: Please write a comment about the purpose of this class. I.e. that it implements sycl queue interface in case dpctl's sycl queue is not available.

File: onedal/common/sycl.cpp (hunk 1 above)
Comment (icfaust) [reply onedal/common/sycl.cpp:39]: I've now left a note @Vika-F please let me know what you think!

File: onedal/common/sycl.cpp (hunk 4 above)
Comment (icfaust) [reply onedal/common/sycl.cpp:31]: Done, let me know what you think

**Comment by icfaust** [comment]
It seems that the previously implemented get_device_id isn't conformant with DPCtl, and checks on filter_string will be commented out in this PR for comparisons.  A follow-up ticket will be made for correcting this, ideally when we do a DLPACK conformance rollout.

File: onedal/common/tests/test_sycl.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+def test_sycl_queue_conversion(queue):
+    if queue is None:
+        pytest.skip("Not a DPCtl queue")
+    SyclQueue = queue.__class__
+    onedal_SyclQueue = _backend.SyclQueue
+    # convert back and forth to test `_get_capsule` attribute
+
+    q = onedal_SyclQueue(queue)
+    # get_device_id must be modified to follow DPCtl conventions
+    # assert q.sycl_device.filter_string in queue.sycl_device.filter_string
```
Comment (ethanglaser) [inline onedal/common/tests/test_sycl.py:75]: is this meant to be commented? test feels incomplete in current state

File: onedal/_device_offload.py [hunk 6]
```diff
@@ -29,37 +29,21 @@
...
     from dpctl.memory import MemoryUSMDevice, as_usm_memory
     from dpctl.tensor import usm_ndarray
+else:
+    import onedal
+
+    # setting fallback to `object` will make if isinstance call
+    # in _get_global_queue always true for situations without the
+    # dpc backend when `device_offload` is used. Instead, it will
+    # fail at the policy check phase yielding a RuntimeError
+    SyclQueue = getattr(onedal._backend, "SyclQueue", object)
```
Comment (ethanglaser) [inline onedal/_device_offload.py:39]: what is object?

File: onedal/common/policy.cpp [hunk 7]
```diff
@@ -41,12 +41,28 @@ void instantiate_default_host_policy(py::module& m) {
 
 #ifdef ONEDAL_DATA_PARALLEL
 
-using data_parallel_policy_t = dal::detail::data_parallel_policy;
+using dp_policy_t = dal::detail::data_parallel_policy;
```
Comment (ethanglaser) [inline onedal/common/policy.cpp:44]: seems redundant if its stated in policy.hpp
```suggestion
```

File: onedal/common/tests/test_sycl.py [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+    # verify copying via the _get_capsule attribute
+    q3 = _backend.SyclQueue(q)
+
+    q_array = [q, q2, q3]
+
+    assert all([queue.sycl_device.has_aspect_fp64 for queue in q_array])
+    assert all([queue.sycl_device.has_aspect_fp16 for queue in q_array])
+    assert all([queue.sycl_device.is_cpu for queue in q_array])
+    assert all([not queue.sycl_device.is_gpu for queue in q_array])
+    assert all(["cpu" in queue.sycl_device.filter_string for queue in q_array])
```
Comment (ethanglaser) [inline onedal/common/tests/test_sycl.py:118]: Have you manually confirmed that results are as expected with Dummy queue?

Comment (icfaust) [reply onedal/common/tests/test_sycl.py:118]: yep! the only thing off is the device_filter and i have a ticket to fix that as it is older and separate functionality

File: onedal/common/sycl_interfaces.cpp [hunk 9]
```diff
@@ -18,19 +18,56 @@
 #include <sycl/sycl.hpp>
 #endif // ONEDAL_DATA_PARALLEL
 
+#include <Python.h>
```
[the remark this answers is not in the document]
Comment (icfaust) [reply onedal/common/sycl_interfaces.cpp:21]: thats right, its the lowest level python c-api header, py::int_ is a basic datatype in pybind11 for PyLong

File: onedal/common/tests/test_sycl.py [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+@pytest.mark.parametrize("device_type", ["cpu", "gpu"])
+@pytest.mark.parametrize("device_number", [None, 0, 1, 2, 3])
+def test_sycl_queue_string_creation(device_type, device_number):
+    # create devices from strings
+    from dpctl import SyclQueue
+    from dpctl._sycl_queue import SyclQueueCreationError
+
+    onedal_SyclQueue = _backend.SyclQueue
+
+    device = ":".join([device_type, str(device_number)]) if device_number else device_type
```
[the remark this answers is not in the document]
Comment (icfaust) [reply onedal/common/tests/test_sycl.py:37]: good call. will switch to device_number is not None

File: onedal/common/tests/test_sycl.py [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+        raised_exception_backend = True
+
+    assert raised_exception_dpctl == raised_exception_backend
+    # get_device_id must be modified to follow DPCtl conventions
+    # this causes filter_string mismatches
+    # if not raised_exception_backend:
+    #    assert (
+    #        onedal_queue.sycl_device.filter_string
+    #        in dpctl_queue.sycl_device.filter_string
+    #    )
```
[the remark this answers is not in the document]
Comment (icfaust) [reply onedal/common/tests/test_sycl.py:59]: yes, there is a follow up ticket to deal with a get_device_id which is outputting incorrect information

File: onedal/common/tests/test_sycl.py (hunk 5 above)
Comment (icfaust) [reply onedal/common/tests/test_sycl.py:75]: there is a ticket for the device filter mismatch, it is currently unusued in our codebase with this PR

File: onedal/_device_offload.py (hunk 6 above)
Comment (icfaust) [reply onedal/_device_offload.py:39]: https://docs.python.org/3/library/functions.html#object the very lowest object in python

File: onedal/common/sycl.cpp (hunk 4 above)
Comment (Vika-F) [reply onedal/common/sycl.cpp:31]: I think having those parts (here and pytorch-related) commented is great!

File: onedal/common/sycl.cpp (hunk 1 above)
Comment (Vika-F) [reply onedal/common/sycl.cpp:39]: I like the comment. It clarifies the things to me.

File: onedal/common/tests/test_sycl.py [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+
+@pytest.mark.skipif(
+    not _is_dpc_backend or not dpctl_available, reason="requires dpc backend and dpctl"
+)
+@pytest.mark.parametrize("queue", get_queues())
+def test_sycl_device_attributes(queue):
+    from dpctl import SyclQueue
+
+    if queue is None:
+        pytest.skip("Not a DPCtl queue")
```
[the remark this answers is not in the document]
Comment (icfaust) [reply onedal/common/tests/test_sycl.py]: Changed.

File: onedal/common/tests/test_sycl.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (icfaust) [reply onedal/common/tests/test_sycl.py]: Done.

**Review by ethanglaser — APPROVED** [review]
